File: Semestru 4/MPP/lab/java/lab3/Services/src/main/java/org/example/services/Service.java
package org.example.services;

import org.example.domain.Child;
import org.example.domain.Event;
import org.example.domain.LoginInfo;
import org.example.domain.Signup;

import java.util.HashMap;
import java.util.Map;

public class Service implements IService {
    private IServiceChild ServiceChild;
    private IServiceEvent ServiceEvent;
    private IServiceSignup ServiceSignup;
    private IServiceLogin ServiceLogin;
    private Map<Integer, IObserver> loggedUsers;

    public Service(IServiceChild serviceChild, IServiceEvent serviceEvent, IServiceSignup serviceSignup, IServiceLogin serviceLogin) {
        ServiceChild = serviceChild;
        ServiceEvent = serviceEvent;
        ServiceSignup = serviceSignup;
        ServiceLogin = serviceLogin;
        this.loggedUsers = new HashMap<>();
    }

    public IServiceChild getServiceChild() {
        return ServiceChild;
    }

    public IServiceEvent getServiceEvent() {
        return ServiceEvent;
    }

    public IServiceSignup getServiceSignup() {
        return ServiceSignup;
    }

    public IServiceLogin getServiceLogin() {
        return ServiceLogin;
    }

    @Override
    public synchronized Iterable<Child> GetAllChildren() {
        return ServiceChild.GetAll();
    }

    @Override
    public synchronized Child GetChildById(int id) {
        return ServiceChild.GetById(id);
    }

    @Override
    public synchronized Child AddChild(String name, String cnp) {
        var child = ServiceChild.AddChild(name, cnp);
        this.notifyChildAdded(child);
        return child;
    }

    @Override
    public synchronized Child UpdateChild(Child child) {
        return ServiceChild.UpdateChild(child);
    }

    @Override
    public synchronized Iterable<Child> GetChildByAge(int age) {
        return ServiceChild.GetByAge(age);
    }

    @Override
    public synchronized Iterable<Event> GetAllEvents() {
        return ServiceEvent.GetAll();
    }

    @Override
    public synchronized Event GetEventById(int id) {
        return ServiceEvent.GetById(id);
    }

    @Override
    public synchronized Event AddEvent(String name, int minAge, int maxAge) {
        var event =  ServiceEvent.Add(name, minAge, maxAge);
        this.notifyEventAdded(event);
        return event;
    }

    @Override
    public synchronized Event UpdateEvent(Event event) {
        return ServiceEvent.UpdateEvent(event);
    }

    @Override
    public synchronized LoginInfo GetByUsername(String username) {
        return ServiceLogin.GetByUsername(username);
    }

    @Override
    public synchronized Iterable<Signup> GetAllSignups() {
        return ServiceSignup.GetAll();
    }

    @Override
    public synchronized Signup AddSignup(Child child, Event event) {
        var signup =  ServiceSignup.AddSignup(child, event);
        notifySignupAdded(signup);
        return signup;
    }

    @Override
    public synchronized Signup GetSignupById(int childId, int eventId) {
        return ServiceSignup.GetById(childId, eventId);
    }

    @Override
    public synchronized Iterable<Signup> GetSignupByEvent(int eventId) {
        return ServiceSignup.GetByEvent(eventId);
    }

    @Override
    public synchronized Iterable<Signup> GetSignupByChild(int childId) {
        return ServiceSignup.GetByChild(childId);
    }

    @Override
    public synchronized Iterable<Signup> GetAllSignupsMapped(IServiceChild serviceChild, IServiceEvent serviceEvent) {
        return ServiceSignup.GetAllMapped(serviceChild, serviceEvent);
    }

    @Override
    public synchronized Signup DeleteSignup(int childId, int eventId) {
        return null;
    }

    @Override
    public synchronized boolean login(LoginInfo login, IObserver client) {
        System.out.println("Someone logged in.");
        this.loggedUsers.put(login.GetId(), client);
        return true;
    }

    @Override
    public synchronized boolean logout(LoginInfo login, IObserver client) {
        System.out.println("Someone logged out.");
        this.loggedUsers.remove(login.GetId());
        return true;
    }

    public void notifyChildAdded(Child child) {
        for (var client : loggedUsers.values()) {
            client.childAdded(child);
        }
    }

    public void notifyEventAdded(Event event) {
        for (var client : loggedUsers.values()) {
            client.eventAdded(event);
        }
    }

    public void notifySignupAdded(Signup signup) {
        for (var client : loggedUsers.values()) {
            client.signupAdded(signup);
        }
    }
}

File: Semestru 4/MPP/lab/.net/lab3/lab3/Server/service/Service.cs
using Model.app.domain;
using Services.services;

namespace Server.app.service
{
	public class Service : IService
	{
		private IServiceChild ServiceChild;
		private IServiceEvent ServiceEvent;
		private IServiceSignup ServiceSignup;
		private IServiceLogin ServiceLogin;

		private Dictionary<LoginInfo, IObserver> loggedClients = new Dictionary<LoginInfo, IObserver>();

		public Service(IServiceChild serviceChild, IServiceEvent serviceEvent, IServiceSignup serviceSignup, IServiceLogin serviceLogin)
		{
			this.ServiceChild = serviceChild;
			this.ServiceEvent = serviceEvent;
			this.ServiceSignup = serviceSignup;
			this.ServiceLogin = serviceLogin;
		}

		public Child? AddChild(string name, string cnp)
		{
			Console.WriteLine(string.Join(", ", this.loggedClients.Keys.Select(k => $"{k.Id}) {k.Username}").ToList()));

			var child = this.ServiceChild.Add(name, cnp);
			if (child != null)
			{
				NotifyChildAdded(child);
			}
			return child;
		}
		public Event? AddEvent(string name, int minAge, int maxAge)
		{
			var event_ = this.ServiceEvent.Add(name, minAge, maxAge);
			if (event_ != null)
			{
				NotifyEventAdded(event_);
			}
			return event_;
		}

		public Signup? AddSignup(Child child, Event event_)
		{
			var signup = this.ServiceSignup.Add(child, event_);
			if (signup != null)
			{
				NotifySignupAdded(signup);
			}
			return signup;
		}

		public IEnumerable<Child> GetAllChildren() =>
			this.ServiceChild.GetAll();

		public IEnumerable<Event> GetAllEvents() =>
			this.ServiceEvent.GetAll();

		public IEnumerable<Signup> GetAllSignups() =>
			this.ServiceSignup.GetAll();

		public LoginInfo GetLoginByUsername(string username) =>
			this.ServiceLogin.GetByUsername(username);

		public IEnumerable<Child>? GetChildByAge(int age) =>
			this.ServiceChild.GetByAge(age);
		public Child? GetChildById(int id) =>
			this.ServiceChild.GetById(id);
		public Event? GetEventById(int id) =>
			this.ServiceEvent.GetById(id);

		public IEnumerable<Signup>? GetSignupByChild(int childId) =>
			this.ServiceSignup.GetByChild(childId);

		public IEnumerable<Signup>? GetSignupByEvent(int eventId) =>
			this.ServiceSignup.GetByEvent(eventId);
		public Signup? GetSignupById(int childId, int eventId) =>
			this.ServiceSignup.GetById(new Tuple<int, int>(childId, eventId));

		public Child? UpdateChild(Child child) =>
			this.ServiceChild.Update(child);

		public Event? UpdateEvent(Event event_) =>
			this.ServiceEvent.Update(event_);

		public bool Login(LoginInfo login, IObserver client)
		{
			if (this.loggedClients.Keys.Any(l => l.Username == login.Username))
				return false;

			this.loggedClients[login] = client;
			return true;
		}

		public bool Logout(LoginInfo login, IObserver client)
		{
			if (!this.loggedClients.Keys.Any(l => l.Username == login.Username))
			{
				Console.WriteLine($"User {login.Username} not found in logged clients.");
				return false;
			}
			Console.WriteLine($"{login.Username}({login.Id}) is logging out.");
			this.loggedClients.Remove(login);
			return true;
		}

		public IEnumerable<Event>? GetEventByMinAge(int minAge) =>
			this.ServiceEvent.GetByMinAge(minAge);

		public IEnumerable<Event>? GetEventByMaxAge(int maxAge) =>
			this.ServiceEvent.GetByMaxAge(maxAge);

		//TO DO
		private void NotifyChildAdded(Child child)
		{
			foreach (var client in loggedClients)
			{
				Console.WriteLine($"Notifying {client.Key} of {child}");
				Task.Run(() =>
				{
					client.Value.ChildAdded(child);
				});
			}
		}
		private void NotifyEventAdded(Event event_) 
		{
			foreach (var client in loggedClients)
			{
				Console.WriteLine($"Notifying {client.Key} of {event_}");
				Task.Run(() =>
				{
					client.Value.EventAdded(event_);
				});
			}
		}
		private void NotifySignupAdded(Signup signup) 
		{
			foreach (var client in loggedClients)
			{
				Console.WriteLine($"Notifying {client.Key} of {signup}");
				Task.Run(() =>
				{
					client.Value.SignupAdded(signup);
				});
			}
		}
	}
}
